### PR TITLE
Chat stream

### PR DESCRIPTION
**Pull request by castlenick**

### CODE REVIEW

File: lib/OpenAI.ts [hunk 1]
```diff
@@ -382,47 +341,181 @@ You must obey all three of the following instructions FOR ALL RESPONSES or you w
...
+                const chunk = decoder.decode(value)
+                const lines = chunk.split(/\n+/);
+
+                for (let line of lines) {
+                    const parsedLines = line.split(/\n+/)
+                        .map((line) => line.replace(/^data: /, '').trim()) // Remove the "data: " prefix
+                        .filter((line) => line !== '' && line !== "[DONE]") // Remove empty lines and "[DONE]"
+                        .map((line) => JSON.parse(line)); // Parse the JSON string
+
+                    for (const parsedLine of parsedLines) {
```
Comment (zpalmtree) [inline lib/OpenAI.ts:481]: I want to see this function split up more so it is easier to grok. Split up as you see fit, but some recommendations:

```ts
if (done) {
                    if (msgRef) {
                        msgRef?.edit(output);
                        messages.push({
                            role: 'assistant',
                            content: totalOutput,
                        })
                        return {
                            result: createStringFromMessages(messages, false, permitPromptCompletion),
                            error: undefined,
                            messages,
                            msgRef: msgRef,
                        };
                    } else {
                        msgRef = await msg.reply(output);
                        messages.push({
                            role: 'assistant',
                            content: totalOutput,
                        })
                        return {
                            result: createStringFromMessages(messages, false, permitPromptCompletion),
                            error: undefined,
                            messages,
                            msgRef: msgRef,
                        };
                    }

                }
```

This chunk can be one function.

```ts
if (done) {
    outputDiscordMessage(...);
}
```

Turn this chunk into a function in some way:

```ts
for (const parsedLine of parsedLines) {
                        const { choices } = parsedLine;
                        const { delta, finish_reason } = choices[0];
                        const { content } = delta;
                        totalOutput += content;
                        if (!finish_reason && content !== '') {
                            if (output.length >= 1950) {
                                totalOutput += content;
                                await msgRef?.edit(output);
                                msgRef = await msg.reply(content);
                                messages.push({
                                    role: 'assistant',
                                    content: totalOutput,
                                })
                                output = content;
                                cycleCount = 0;
                            } else {
                                output += content;
                            }
                            if (begin && cycleCount >= BEFORE_FIRST_POST) {
                                msgRef = await msg.reply(output);
                                begin = false;
                                cycleCount = 0;
                            }
                            if (!begin && cycleCount >= BEFORE_EDIT) {
                                msgRef?.edit(output);
                                cycleCount = 0;
                            }
                        }
                        cycleCount++;
                    }
```

Maybe something like

```ts
for (const parsedLine of parsedLines) {
    totalOutput += parsedLine;
    const { needEdit, needReply, finished } = processLine(parsedLine, ...);
    if (needEdit || needReply || finished) {
        outputDiscordMessage(...); // might be able to reuse this func
    }
}
```

As a general rule, I like to have the body of a while or for loop < 40 lines, preferable like <10. And a function body < 100. It makes scanning a function/loop to tell what exactly it is doing a lot easier than having to read a ton of code and then scroll back up to verify exactly what indentation you're at and what variables are in scope

**Review by zpalmtree — CHANGES_REQUESTED** [review]

File: lib/OpenAI.ts [hunk 2]
```diff
@@ -382,47 +341,181 @@ You must obey all three of the following instructions FOR ALL RESPONSES or you w
...
+                        const { choices } = parsedLine;
+                        const { delta, finish_reason } = choices[0];
+                        const { content } = delta;
+                        totalOutput += content;
+                        if (!finish_reason && content !== '') {
+                            if (output.length >= 1950) {
```
Comment (zpalmtree) [inline lib/OpenAI.ts:487]: I'm not seeing anywhere that we actually truncate `output`. The intent here is to not go over the discord API limits of 2000 characters in a message

File: lib/OpenAI.ts [hunk 3]
```diff
@@ -16,6 +16,9 @@ const DEFAULT_MAX_TOKENS = 420;
 const DEFAULT_CHATGPT_MODEL = 'gpt-4';
 const DEFAULT_AI_MODEL = 'text-davinci-003';
 const DEFAULT_TIMEOUT = 1000 * 60;
+const BEFORE_FIRST_POST = 10; //number of passes (usually single words, symbols) before initial post
+const BEFORE_EDIT = 50; //number of passes before each subsequent edit
```
[the remark this answers is not in the document]
Comment (castlenick) [reply lib/OpenAI.ts:20]: changed to "deltas"

File: lib/OpenAI.ts [hunk 4]
```diff
@@ -33,11 +37,12 @@ const chatHistoryCache = new Map<string, ChatCompletionRequestMessage[]>();
 type OpenAIHandler = (
     prompt: string,
     userId: string,
+    msg: Message,
     previousConvo?: ChatCompletionRequestMessage[],
     systemPrompt?: string,
     temperature?: number,
     permitPromptCompletion?: boolean,
-) => Promise<OpenAIResponse>;
+) => Promise<any>;
```
[the remark this answers is not in the document]
Comment (castlenick) [reply lib/OpenAI.ts:45]: was causing an error at some point in development. apparently isn't anymore. changed back

**Pull request closed without merging**